doc/man/gnunet-transport-check.1
.TH GNUNET-TRANSPORT-CHECK "1" "19 Dec 2003" "GNUnet"
.SH NAME
gnunet-transport-check \- a tool to test a GNUnet transport service
.SH SYNOPSIS
.B gnunet\-transport\-check
[\fIOPTIONS\fR] 
.SH DESCRIPTION
.PP
gnunet-transport-check can be used to test or profile
a GNUnet transport service.  The tool can be used to test
both the correctness of the software as well as the correctness
of the configuration.  gnunet-transport-check features two modes,
called loopback mode and ping mode.  In loopback mode the test is limited to testing if the
transport can be used to communicate with itself (loopback).
This mode does not include communication with other peers which
may be blocked by firewalls and other general Internet connectivity
problems.  The loopback mode is particularly useful to test
the SMTP transport service since this service is fairly hard to
configure correctly and most problems can be reveiled by just
testing the loopback.  In ping mode the tool will attempt to download
peer advertisements from the URL specified in the configuration file
and then try to contact each of the peers.  Note that it is perfectly
normal that some peers do not respond, but if no peer responds something
is likely to be wrong.  The configuration is always taken
from the configuration file.  Do not run gnunetd while running
gnunet-transport-check since the transport services can not 
be used by two processes at the same time.
.PP
gnunet-transport-check will always produce an error-message for
the NAT transport in loopback mode.  If NAT is configured in accept-mode (as in,
accept connections from peers using network address translation),
the check will fail with the message "could not create HELO",
which is correct since the peer itself is clearly not going to
advertise itself as a NAT.  If the peer is configured in NAT-mode,
that is, the peer is behind a NAT box, the message will be
'could not connect'.  For NAT, both messages are NOT errors
but exactly what is supposed to happen.
.PP
Similarly, a NAT-ed peer should typically configure the TCP transport
to use port 0 (not listen on any port).  In this case, 
gnunet-transport-check will print 'could not create HELO' for the
TCP transport.  This is also ok.  In fact, a correctly configured
peer using NAT should give just two errors (could not connect for
tcp and could not create HELO for NAT) when tested using
gnunet-transport-check.  The reason is, that gnunet-transport-check
only tests loopback connectivity, and for a NAT-ed peer, that just
does not apply.
.PP
Note that in ping mode the HTTP download times out after 5 minutes,
so if the list of peers is very large and not all peers can be
queried within the 5 minutes the tool may abort before trying all
peers.
.TP
\fB\-c \fIFILENAME\fR, \fB\-\-config=\fIFILENAME\fR
use config file (default: /etc/gnunet.conf)
.TP
\fB\-h\fR, \fB\-\-help\fR
print help page
.TP
\fB\-L \fILOGLEVEL\fR, \fB\-\-loglevel=\fILOGLEVEL\fR
change the loglevel. Possible values for \fILOGLEVEL\fR are NOTHING, FATAL, ERROR, FAILURE, WARNING, MESSAGE, INFO, DEBUG, CRON and EVERYTHING.
.TP
\fB\-p\fR, \fB\-\-ping\fR
use ping mode (loopback mode is default)
.TP
\fB\-r\fI COUNT \fB\-\-repeat=\fICOUNT\fR
send COUNT messages in a sequence over the same connection
.TP
\fB\-s\fI SIZE \fB\-\-size=\fISIZE\fR
test using the specified message size, default is 11
.TP
\fB\-t\fI TRANSPORT\fR, \fB\-\-transport=\fITRANSPORT\fR
run using the specified transport, if not given the transports
configured in the configuration file are used.
.TP
\fB\-v\fR, \fB\-\-version\fR
print the version number
.TP
\fB\-V\fR, \fB\-\-verbose\fR
be verbose
.SH NOTES
gnunet-transport-check can run for a long time, depending on
how high you have set the \fICOUNT\fR level. Run first with small numbers
for \fICOUNT\fR to get an initial estimate on the runtime.
.SH FILES
.TP
/etc/gnunet.conf
default gnunetd configuration file
.SH "REPORTING BUGS"
Report bugs by using mantis <http://www.ovmj.org/~mantis/> or by sending electronic mail to <gnunet-developers@gnu.org>
.SH "SEE ALSO"
\fBgnunet.conf\fP(5), \fBgnunetd\fP(1)


ChangeLog
Mon Feb 28 00:33:03 EST 2005
	Releasing GNUnet 0.6.6b.

Sun Feb 27 19:27:44 EST 2005
	Fixing mantis 839.

Sun Feb 27 18:44:11 EST 2005
	Fixing mantis 833/834.

Sun Feb 27 18:15:02 EST 2005
	Fixing serious problem with LE in gnunet-insert.
	Fixing mantis 828.

Thu Feb 17 01:14:30 EST 2005
	Various memory leaks for Win32 fixed.
	Added support for GNUNET(D)_CONFIG environment variable.

Wed Feb  9 13:33:38 EST 2005
        Fixed problems with the logrotate code.  Releasing GNUnet 0.6.6a.

Sun Jan 11 19:58:13 CET 2005
	Releasing GNUnet 0.6.6

Sun Jan 11 19:24:27 CET 2005
        gnunet-gtk now resumes aborted downloads automatically.

Sun Nov 14 16:49:02 EST 2004
	Debugged collections, seems to work (not that it looks pretty,
	but it basically works, the rest is UI work).  Fixed deadlock
	in gnunet-gtk logging.  Releasing GNUnet 0.6.5.

Thu Nov  4 09:04:48 EST 2004
	Added (highly experimental) support for collections.

Sat Sep 25 22:29:48 EST 2004
	Releasing GNUnet 0.6.4a.

Sun Sep 19 23:39:04 EST 2004
	Various code cleanups and minor bugfixes.

Sun Aug 29 15:11:28 IST 2004
	Releasing GNUnet 0.6.4.

Tue Aug 24 20:59:14 IST 2004
	Significant enhancements to namespaces.  A namespace can
	now be annotated with meta-data at the time that it is
	created.
	
Sat Aug 21 01:00:06 IST 2004
	Adding GNU gettext support to configure.
	
Thu Aug 19 01:26:55 IST 2004
	Migrating gnunet-gtk to GTK 2.0 (tested with 2.4).

Mon Aug 16 00:30:59 IST 2004
	Removed gcry.  Added dependency on GNU MP library (libgmp)
	for kblocks.

Tue Aug  3 17:57:43 IST 2004
	Fixed bugs in gnunet-update.  Fixed segfault in AFS startup.
	Fixed warnings in gnunet-peer-info/gnunetd.  Fixed packaging
	error (not all of util/gcry included in distribution).
	Releasing GNUnet 0.6.3a.

Mon Aug  2 20:06:07 IST 2004
	Releasing GNUnet 0.6.3.

Sun Aug  1 01:31:19 IST 2004
	Changed most places from HexName to the new, shorter EncName.
	Pushed URIs for namespaces and keyword searches into more
	widespread use.  Replaced all occurences of sprintf with
	SNPRINTF (to obtain extra bounds check).
	
Sun Jul 11 13:54:01 IST 2004
	Moved testcases for libgnunet_util to util/ directory.
	Made DB testcases in afs/module/ run for each available
	database and not just the one configured in /etc/gnunet.conf.

Fri Jun 25 21:05:59 IST 2004
	Releasing GNUnet 0.6.2b (resolves various minor bugs).

Wed May  5 17:34:39 EST 2004
	Releasing GNUnet 0.6.2a (resolves libgcrypt and packageing
	issues, improved gtk-statistics, other minor bugfixes).
	
Fri Apr 30 21:36:23 EST 2004
	Fixed synchronization problem in cron (only relevant for
	parallel downloading).  Releasing GNUnet 0.6.2.

Mon Apr 26 21:13:45 EST 2004
	Added support to gnunet-check to convert file index database
	to new design.  Removed support for pre 0.6.1b bloomfilter
	conversion (would have been too messy to support both).
	Added old-version check to gnunetd startup to make updates
	more systematic in the future (not perfect, though).

Mon Apr 26 19:27:29 EST 2004
	Recursive insertion of directories with gnunet-gtk 
	now works including the progress bar.

Sun Apr 25 18:24:55 EST 2004
	Global changes to fix bug #698.  The fix is still 
	incomplete with respect to pthread.h specific structs.
	
Sun Apr 25 15:20:29 EST 2004
	Fixed compile error in IPv6 transports.

Fri Apr 23 23:38:01 EST 2004
	Added recursive insertion of directories to gnunet-gtk.
	Progress-bar does not work yet, also libextractor is
	always used (no option to turn it off).
	
Fri Apr 23 21:49:18 EST 2004
	Removed need to specify _protocol in configuration file
	(shorter is better).

Wed Apr 21 23:32:36 EST 2004
	Added code to bound priority used by clients by twice the
	average priority of requests from other peers that are 
	routed at the moment.  That prevents clients from exposing
	themselves with excessively high priorities (and also 
	avoids giving too much credit/trust needlessly).

Tue Apr 20 22:27:19 EST 2004
	Added possibility of using a link instead of making a copy
	if gnunetd and the insert-client both run on the same machine.

Tue Apr 20 00:00:58 EST 2004
	Added network transparency for indexing files (by copying
	them to a GNUnet(d) specific directory).  This breaks
	unindexing (for files indexed with previous versions) and
	forces the user to make a copy of the file.  In the near
	future we should resolve this problem by allowing the use
	of a link.

Sun Apr 18 17:24:34 EST 2004
	Adding recursive, parallel download of directories to
	gnunet-download

Sun Apr 18 01:57:32 EST 2004
	gnunet-transport-check can now be used to check
	external connectivity (with other peers).
	
Sat Apr 17 17:46:19 EST 2004
	Reduced memory consuption for AFS insertion and
	download.
	
Sat Apr 10 21:17:46 EST 2004
	Implemented statistics-plot in gnunet-gtk based on
	gnome-system-monitor code.

Fri Apr  9 22:01:51 EST 2004
	Added support for libextractor > 0.2.7.
	
Fri Apr  9  0:29:20 EST 2004
	Completed --help conversion.

Wed Mar 31 15:40:51 EST 2004
	Releasing 0.6.1d.

Tue Mar 30 22:42:13 EST 2004
	Integrated gnunet-pseudonym-create, -delete and -list into
	gnunet-pseudonym as well as gnunet-directory-list, -emptydb
	and -print into gnunet-directory.

Tue Mar  2 10:46:01 EST 2004
	Fixing bug preventing download for indexed files <= 1k.

Sat Feb 28 03:43:34 EST 2004
	Releasing 0.6.1c.

Sat Feb 21 06:16:47 EST 2004
	Modified requestmanager to improve performance at the end of the 
	download.

Thu Feb 19 00:48:33 EST 2004
	Fixed integer underflow possibly causing slow download performance.
	
Sun Feb 15 15:17:05 EST 2004
	Fixed weak key failures with libgcrypt.

Sun Feb 15 13:14:44 EST 2004
	New routing code (untested) commited.

Sun Feb 15 11:58:10 EST 2004
	Fixing problem with 64-bit size_t on alpha (#655).

Sat Jan 31 04:32:33 EST 2004
	Releasing 0.6.1b.

Fri Jan 23 00:00:09 EST 2004
	Fixed indexed content being overwritten by migrated content
	
Thu Jan 22 19:27:33 EST 2004
	Fixed integer-overflow problems in statuscalls.

src/server/gnunet-transport-check.c
/*
     This file is part of GNUnet.
     (C) 2001, 2002, 2003, 2004 Christian Grothoff (and other contributing authors)

     GNUnet is free software; you can redistribute it and/or modify
     it under the terms of the GNU General Public License as published
     by the Free Software Foundation; either version 2, or (at your
     option) any later version.

     GNUnet is distributed in the hope that it will be useful, but
     WITHOUT ANY WARRANTY; without even the implied warranty of
     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
     General Public License for more details.

     You should have received a copy of the GNU General Public License
     along with GNUnet; see the file COPYING.  If not, write to the
     Free Software Foundation, Inc., 59 Temple Place - Suite 330,
     Boston, MA 02111-1307, USA.
*/

/**
 * @file server/gnunet-transport-check.c
 * @brief Test for the transports.  
 * @author Christian Grothoff
 *
 * This utility can be used to test if a transport mechanism for
 * GNUnet is properly configured.
 */

#include "gnunet_util.h"
#include "transport.h"
#include "httphelo.h"
#include "keyservice.h"
#include "knownhosts.h"

#define DEBUG_TRANSPORT_CHECK NO

#define TEST_CRC 25116123
#define DEFAULT_MSG "Hello World"

static char * testmsg;

static MessagePack * receit;
static CoreAPIForTransport capi;
static Semaphore * sem;
static int terminate;
static cron_t timeout = 15 * cronSECONDS;

static void receive(MessagePack * mp) {
  if ( (sem == NULL) || (receit != NULL) ) {
    FREE(mp->msg);
    FREE(mp);
    return; /* spurious receive or double-receive, happens */
  }
  if (testConfigurationString("GNUNET-TRANSPORT-CHECK",
			      "VERBOSE",
			      "YES")) 
    fprintf(stderr, ".");
  receit = mp;
  SEMAPHORE_UP(sem);
}

static void semUp(Semaphore * sem) {
  terminate = YES;
  SEMAPHORE_UP(sem);
}

/**
 * Test the given transport API.
 */
static void testTAPI(TransportAPI * tapi,
		     int * res) {
  HELO_Message * helo;
  TSession * tsession;
  unsigned int repeat;
  cron_t start;
  cron_t end;

  if (tapi == NULL)
    errexit("Could not initialize transport!\n");

  helo = NULL;
  if (OK != tapi->startTransportServer()) {
    fprintf(stderr,
	    _("Could not start transport server.\n"));
    *res = SYSERR;
    return;
  }
  if (OK != tapi->createHELO(&helo)) {
    fprintf(stderr,
	    _("Could not create HELO.\n"));
    tapi->stopTransportServer();
    *res = SYSERR;
    return;
  }
  tsession = NULL;
  if (OK != tapi->connect(helo,
			  &tsession)) {
    fprintf(stderr,
	    _("Could not connect.\n"));
    *res = SYSERR;
    tapi->stopTransportServer();
    FREE(helo);
    return;
  }
  repeat = getConfigurationInt("TRANSPORT-CHECK",
			       "REPEAT");
  if (repeat == 0) {
    repeat = 1;
    setConfigurationInt("TRANSPORT-CHECK",
			"REPEAT",
			1);
  }
  sem = SEMAPHORE_NEW(0);
  cronTime(&start);
  while (repeat > 0) {
    repeat--;
    receit = NULL;
    if (OK != tapi->send(tsession,
			 testmsg,
			 strlen(testmsg),
			 NO,
			 TEST_CRC)) {
      fprintf(stderr,
	      _("Could not send.\n"));
      *res = SYSERR;
      tapi->disconnect(tsession);
      tapi->stopTransportServer();
      SEMAPHORE_FREE(sem);
      return;
    }
    addCronJob((CronJob)&semUp,
	       timeout,
	       0,
	       sem);
    SEMAPHORE_DOWN(sem);
    suspendCron();
    delCronJob((CronJob)&semUp,
	       0,
	       sem); 
    resumeCron();
    if (receit == NULL) {
      fprintf(stderr,
	      _("Did not receive message within %llu ms.\n"),
	      timeout);
      *res = SYSERR;
      tapi->disconnect(tsession);
      tapi->stopTransportServer();
      SEMAPHORE_FREE(sem);
      return;
    }
    if ( (receit->crc != TEST_CRC) ||
	 (receit->size != strlen(testmsg)) ||
	 (receit->isEncrypted != NO) ||
	 (0 != memcmp(capi.myIdentity,
		      &receit->sender,
		      sizeof(HostIdentity))) ||
	 (0 != memcmp(receit->msg,
		      testmsg,
		      strlen(testmsg))) ) {
      fprintf(stderr,
	      _("Message received was invalid.\n"));
      *res = SYSERR;
      tapi->disconnect(tsession);
      tapi->stopTransportServer();
      SEMAPHORE_FREE(sem);
      return;
    }
    FREE(receit->msg);
    FREE(receit);
  }
  cronTime(&end);
  if (OK != tapi->disconnect(tsession)) {
    fprintf(stderr,
	    _("Could not disconnect.\n"));
    *res = SYSERR;
    tapi->stopTransportServer();
    SEMAPHORE_FREE(sem);
    return;
  }
  if (OK != tapi->stopTransportServer()) {
    fprintf(stderr,
	    _("Could not stop server.\n"));
    *res = SYSERR;
    SEMAPHORE_FREE(sem);
    return;
  }

  SEMAPHORE_FREE(sem);
  printf(_("Transport OK.  It took %ums to transmit %d messages of %d bytes each.\n"),
	 (unsigned int) ((end - start)/cronMILLIS),  
	 getConfigurationInt("TRANSPORT-CHECK",
			     "REPEAT"),
	 strlen(testmsg));
}


static void testPING(HELO_Message * xhelo,
		     int * stats) {  
  TSession * tsession;
  PINGPONG_Message pmsg;
  HELO_Message * helo;
  HELO_Message * myHelo;
  char * msg;
  int len;
  int again;
  int reply;
  
  if (testConfigurationString("GNUNET-TRANSPORT-CHECK",
			      "VERBOSE",
			      "YES")) {
    char * str;
    str = heloToString(xhelo);
    fprintf(stderr, 
	    _("\nContacting '%s'."),
	    str);
    FREE(str);
  } else
    fprintf(stderr, ".");
  helo = MALLOC(ntohs(xhelo->header.size));
  memcpy(helo, xhelo, ntohs(xhelo->header.size));

  stats[0]++; /* one more seen */
  if (NO == isTransportAvailable(ntohs(helo->protocol))) {
    fprintf(stderr, 
	    " Transport %d not available\n",
	    ntohs(helo->protocol));
    FREE(helo);
    return;
  }
  myHelo = NULL;
  if (OK != transportCreateHELO(ntohs(xhelo->protocol),
				&myHelo)) {
    FREE(helo);
    return;
  }
  if (testConfigurationString("GNUNET-TRANSPORT-CHECK",
			      "VERBOSE",
			      "YES")) 
    fprintf(stderr, ".");

  stats[1]++; /* one more with transport 'available' */

  pmsg.header.size = htons(sizeof(PINGPONG_Message));
  pmsg.header.requestType = htons(p2p_PROTO_PING);
  memcpy(&pmsg.receiver,
	 &helo->senderIdentity,
	 sizeof(HostIdentity));
  pmsg.challenge = rand();
 
  
  tsession = NULL;
  if (OK != transportConnect(helo, 
			     &tsession)) {
    FREE(helo);
    fprintf(stderr, 
	    " Connection failed\n");
    return;
  }
  if (testConfigurationString("GNUNET-TRANSPORT-CHECK",
			      "VERBOSE",
			      "YES")) 
    fprintf(stderr, ".");
  
  sem = SEMAPHORE_NEW(0);
  len = sizeof(PINGPONG_Message) + ntohs(myHelo->header.size);
  msg = MALLOC(len);
  memcpy(msg,
	 myHelo,
	 ntohs(myHelo->header.size));
  memcpy(&msg[ntohs(myHelo->header.size)],
	 &pmsg,
	 sizeof(PINGPONG_Message));
  FREE(myHelo);
  /* send ping */
  if (OK != transportSend(tsession,
			  msg,
			  len,
			  NO,
			  crc32N(msg, len))) {
    fprintf(stderr, 
	    " Send failed.\n");
    FREE(msg);
    transportDisconnect(tsession);
    return;
  }
  FREE(msg);
  if (testConfigurationString("GNUNET-TRANSPORT-CHECK",
			      "VERBOSE",
			      "YES")) 
    fprintf(stderr, ".");
  /* check: received pong? */
  terminate = NO;
  addCronJob((CronJob)&semUp,
	     timeout,
	     5 * cronSECONDS,
	     sem);
  reply = 0;
  again = 1;
  while (again && (terminate == NO)) {
    again = 0;
    SEMAPHORE_DOWN(sem);
    if (receit != NULL) {
      p2p_HEADER * part;
      PINGPONG_Message * pong;      
      unsigned int pos;
      unsigned short plen;
      
      again = 1;
      if (0 != memcmp(&receit->sender,
		      &xhelo->senderIdentity,
		      sizeof(HostIdentity))) {
#if DEBUG_TRANSPORT_CHECK
	EncName enc1, enc2;
	hash2enc(&xhelo->senderIdentity.hashPubKey,
		 &enc1);
	hash2enc(&receit->sender.hashPubKey,
		 &enc2);
	fprintf(stderr, 
		"%s!=%s",
		(char*)&enc1, 
		(char*)&enc2);
#endif
	FREE(receit->msg);
	FREE(receit);
	receit = NULL;
	continue; /* different peer, ignore */
      }
      
      reply = 1;
      pos = 0;
      while (pos < receit->size) {
	part = (p2p_HEADER*) &((char*)receit->msg)[pos];
	plen = ntohs(part->size);
#if DEBUG_TRANSPORT_CHECK 
	fprintf(stderr, "PRT<%d,%d>:%d@%d",
		plen,
		ntohs(part->requestType),
		pos,
		receit->size);
#endif
	pos += plen;
	if ( (pos > receit->size) || (plen < sizeof(p2p_HEADER)) ) {
	  fprintf(stderr,
		  "!F");
	  break; /* malformed */
	}

	if ( (plen == sizeof(PINGPONG_Message)) &&
	     (part->requestType == htons(p2p_PROTO_PONG)) ) {	  
	  pong = (PINGPONG_Message*) part;
	  pong->header.requestType = htons(p2p_PROTO_PING);
	  if (0 == memcmp(&pmsg,
			  pong,
			  sizeof(PINGPONG_Message))) {
	    if (testConfigurationString("GNUNET-TRANSPORT-CHECK",
					"VERBOSE",
				      "YES")) 
	      fprintf(stderr, _("OK!"));
	    stats[2]++;
	    reply = 2;
	    again = 0;
	    break;
	  } else {
	    if (testConfigurationString("GNUNET-TRANSPORT-CHECK",
					"VERBOSE",
					"YES")) 
	      fprintf(stderr,
		      "!"); /* invalid pong */
	  }
	} 
      }
      FREE(receit->msg);
      FREE(receit);
      receit = NULL;    
    }
  }
  if (testConfigurationString("GNUNET-TRANSPORT-CHECK",
			      "VERBOSE",
			      "YES")) {
    if (reply == 1)
      fprintf(stderr, _("No PONG received.\n"));
    else if (reply == 0)
      fprintf(stderr, 
	      _("No reply received within %llums.\n"),
	      timeout);
  }
  suspendCron();
  delCronJob((CronJob)&semUp,
	     5 * cronSECONDS,
	     sem); 
  resumeCron();
  SEMAPHORE_FREE(sem);
  sem = NULL;
  if (receit != NULL) {
    FREE(receit->msg);
    FREE(receit);
    receit = NULL;        
  }
  transportDisconnect(tsession);  
}

/* dead code to make linker happy without
   dragging in heloexchange.c */
int receivedHELO(p2p_HEADER * message) {
  return OK;
}

/**
 * Perform option parsing from the command line. 
 */
static int parser(int argc, 
		  char * argv[]) {
  int cont = OK;
  int c;

  /* set the 'magic' code that indicates that
     this process is 'gnunetd' (and not any of
     the user-tools).  Needed such that we use
     the right configuration file... */
  FREENONNULL(setConfigurationString("GNUNETD",
				     "_MAGIC_",
				     "YES"));

  FREENONNULL(setConfigurationString("GNUNETD",
				     "LOGFILE",
				     NULL));
  while (1) {
    int option_index = 0;
    static struct GNoption long_options[] = {
      { "loglevel",1, 0, 'L' },
      { "config",  1, 0, 'c' },
      { "version", 0, 0, 'v' },
      { "help",    0, 0, 'h' },
      { "transport", 1, 0, 't' },
      { "repeat",  1, 0, 'r' },
      { "size",    1, 0, 's'},
      { "Xrepeat", 1, 0, 'x' },
      { "timeout", 1, 0, 'T' },
      { "ping",    0, 0, 'p' },
      { "Xport",   1, 0, 'P' },
      { "verbose", 0, 0, 'V' },
      { 0,0,0,0 }
    };
    
    c = GNgetopt_long(argc,
		      argv, 
		      "vhc:L:t:r:s:X:P:pVT:", 
		      long_options, 
		      &option_index);
    
    if (c == -1) 
      break;  /* No more flags to process */
    
    switch(c) {
    case 'p': 
      FREENONNULL(setConfigurationString("TRANSPORT-CHECK",
					 "PING",
					 "YES"));
      break;
    case 'P':{
      unsigned int port;
      if (1 != sscanf(GNoptarg, "%ud", &port)) {
	LOG(LOG_FAILURE, 
	    "You must pass a number to the -P option.\n");
	return SYSERR;
      } else {
	setConfigurationInt("TCP", "PORT", port);
	setConfigurationInt("UDP", "PORT", port);
	setConfigurationInt("TCP6", "PORT", port);
	setConfigurationInt("UDP6", "PORT", port);
	setConfigurationInt("HTTP", "PORT", port);
      }
      break;
    }
    case 's':{
      unsigned int size;
      if (1 != sscanf(GNoptarg, "%ud", &size)) {
	LOG(LOG_FAILURE, 
	    _("You must pass a number to the '%s' option.\n"),
	    "-s");
	return SYSERR;
      } else {
	if (size == 0)
	  size = 2;
	else
	  size++;
	testmsg = MALLOC(size);
	testmsg[--size] = '\0';
	while (size > 0)
	  testmsg[--size] = 'A';
      }
      break;
    }
    case 'r':{
      unsigned int repeat;
      if (1 != sscanf(GNoptarg, "%ud", &repeat)) {
	LOG(LOG_FAILURE, 
	    _("You must pass a number to the '%s' option.\n"),
	    "-r");
	return SYSERR;
      } else {
	setConfigurationInt("TRANSPORT-CHECK",
			    "REPEAT",
			    repeat);
      }
      break;
    }
    case 'X':{
      unsigned int repeat;
      if (1 != sscanf(GNoptarg, "%ud", &repeat)) {
	LOG(LOG_FAILURE, 
	    _("You must pass a number to the '%s' option.\n"),
	    "-X");
	return SYSERR;
      } else {
	setConfigurationInt("TRANSPORT-CHECK",
			    "X-REPEAT",
			    repeat);
      }
      break;
    }
    case 'T':{
      if (1 != sscanf(GNoptarg, "%llu", &timeout)) {
	LOG(LOG_FAILURE, 
	    _("You must pass a number to the '%s' option.\n"),
	    "-T");
	return SYSERR;
      }
      break;
    }
    case 'c':
      FREENONNULL(setConfigurationString("FILES",
					 "gnunet.conf",
					 GNoptarg));
      break;
    case 't':
      FREENONNULL(setConfigurationString("GNUNETD",
					 "TRANSPORTS",
					 GNoptarg));
      break;
    case 'v': 
      printf("gnunet-transport-check v%s\n",
	     VERSION);
      cont = SYSERR;
      break;
    case 'V':
      FREENONNULL(setConfigurationString("GNUNET-TRANSPORT-CHECK",
					 "VERBOSE",
					 "YES"));
      break;
    case 'h': {
      static Help help[] = {
	HELP_CONFIG,
	HELP_HELP,
	HELP_LOGLEVEL,
	{ 'p', "ping", NULL,
	  gettext_noop("ping peers from HOSTLISTURL that match transports") },
	{ 'r', "repeat", "COUNT",
	  gettext_noop("send COUNT messages") },
	{ 's', "size", "SIZE",
	  gettext_noop("send messages with SIZE bytes payload") },
	{ 't', "transport", "TRANSPORT",
	  gettext_noop("specifies which TRANSPORT should be tested") },
	{ 'T', "timeout", "MS",
	  gettext_noop("specifies after how many MS to time-out") },
	HELP_VERSION,
        HELP_VERBOSE,
	HELP_END,
      };
      formatHelp("gnunet-transport-check [OPTIONS]",
		 _("Tool to test if GNUnet transport services are operational."),
		 help);  
      cont = SYSERR;
      break;
    }
    case 'L':
      FREENONNULL(setConfigurationString("GNUNETD",
					 "LOGLEVEL",
					 GNoptarg));
      break;
    default:
      LOG(LOG_FAILURE, 
	  _("Use --help to get a list of options.\n"));
      cont = SYSERR;    
    } /* end of parsing commandline */
  }
  if (GNoptind < argc) {
    LOG(LOG_WARNING, 
	_("Invalid arguments: "));
    while (GNoptind < argc)
      LOG(LOG_WARNING, 
	  "%s ", argv[GNoptind++]);
    LOG(LOG_FATAL,
	_("Invalid arguments. Exiting.\n"));
    return SYSERR;
  }
  return cont;
}

CoreAPIForTransport * getCoreAPIForTransport() {
  return &capi;
}

int main(int argc, char *argv[]) {
  int res;
  int Xrepeat;
  char * trans;
  int ping;
  char * url;
  int i;
  int stats[3];

  if (OK != initUtil(argc, argv, &parser)) {
    return SYSERR;
  }
  if (testmsg == NULL)
    testmsg = STRDUP(DEFAULT_MSG);

  trans = getConfigurationString("GNUNETD",
				 "TRANSPORTS");
  if (trans == NULL)
    errexit(_("You must specify a non-empty set of transports to test!\n"));
  ping = testConfigurationString("TRANSPORT-CHECK",
				 "PING",
				 "YES");
  if (! ping)
    printf(_("Testing transport(s) %s\n"),
	   trans);
  else
    printf(_("Available transport(s): %s\n"),
	   trans);
  FREE(trans);
  if (! ping) {
    /* disable blacklists (loopback is often blacklisted)... */
    FREENONNULL(setConfigurationString("TCP",
				       "BLACKLIST",
				       NULL));
    FREENONNULL(setConfigurationString("UDP",
				       "BLACKLIST",
				       NULL));
    FREENONNULL(setConfigurationString("TCP6",
				       "BLACKLIST",
				       NULL));
    FREENONNULL(setConfigurationString("UDP6",
				       "BLACKLIST",
				       NULL));
    FREENONNULL(setConfigurationString("HTTP",
				       "BLACKLIST",
				       NULL));
  }
  initKnownhosts();
  initTransports();
  if (ping) {
    initKeyService("gnunet-transport-check");
  } 
  startCron();

  capi.version = 0;
  capi.receive = &receive;
  capi.myIdentity = &myIdentity;
  Xrepeat = getConfigurationInt("TRANSPORT-CHECK",
				"X-REPEAT");
  if (Xrepeat == 0)
    Xrepeat = 1;
  res = OK;
  if (ping) {
    initHttpHelo();
    startTransports();

    stats[0] = 0;
    stats[1] = 0;
    stats[2] = 0;
    url = getConfigurationString("GNUNETD",
				 "HOSTLISTURL");
    if (url != NULL) {
      i = strlen(url);
      while (i > 0) {
	i--;
	if (url[i] == ' ') {
#if DEBUG_TRANSPORT_CHECK
	  fprintf(stderr,
		  "URL: %s\n", 
		  &url[i+1]);
#endif
	  downloadHostlistHelper(&url[i+1],
				 (HELO_Callback)&testPING,
				 &stats[0]);
	  url[i] = '\0';
	}
      } 
#if DEBUG_TRANSPORT_CHECK
      fprintf(stderr,
	      "URL: %s\n",
	      &url[0]);
#endif
      downloadHostlistHelper(&url[0],
			     (HELO_Callback)&testPING,
			     &stats[0]);
      FREE(url);
      fprintf(stderr, "\n");
    } else {
      printf(_("No URL specified in configuration section '%s' under '%s'!\n"),
	     "GNUNETD", "HOSTLISTURL");
    }
    printf(_("%d out of %d peers contacted successfully (%d times transport unavailable).\n"),
	   stats[2],
	   stats[1],
	   stats[0] - stats[1]);
    doneHttpHelo();
    stopTransports();
  } else {
    while (Xrepeat-- > 0) 
      forEachTransport((TransportCallback)&testTAPI,
		       &res);  
  }
  stopCron();
  doneTransports();
  if (ping) 
    doneKeyService();
  doneKnownhosts();
  FREE(testmsg);
  doneUtil();
  if (res == OK)
    return 0;
  else
    return -1;
}


/* end of gnunet-transport-check */
